### DISCUSSION TITLE
Some help with injecting styles

### DISCUSSION THREAD

**Discussion by lewis-fidlers** [vikejs/vike#1804] [Discussion]
Hi, 

I already posted on an issue which i think is somewhat related. (https://github.com/vikejs/vike-vue/issues/141) But this appears to be the better place to ask the question. :)

We are currently starting to move over from a very old stack and are heavily tied to [ant.design](https://ant.design/) 
We currently have vike set up with vike-react. And are using ant.design for our ui stuff.

As per their guide, i should be able to extract the css and inject it into the page. [their ssr docs](https://ant-design.github.io/antd-style/guide/ssr). I believe having my own [onRender](https://vike.dev/onRenderHtml) should do the trick. Though i would lose whatever vike-react has per default, which i would like to keep as well. 

So, I'm having a hard time figuring out how this should be achieved in combination with vike-react, any pointers on what the best approach would be are welcomed :)

Cheers, Lewis

**Comment by brillout** [comment]
Thanks for reaching out and, yes, being able to integrate tools such as antd with vike-react is very much a priority. Can you publish a minimal example of using antd with [Vike but without vike-react](https://vike.dev/new#without-vike-react-vue-solid)? I'll then accordingly create hooks on vike-react's side.

**Comment by lewis-fidlers** [comment]
Thanks for the response, it is highly appreciated! :) 
### Example and explanation

Example can be found [here](https://github.com/lewis-fidlers/vike-ant-design-minimal)

In theory, not a lot is needed to get it working. (from what i can tell anyway, as mentioned I'm pretty new to this.) 

I had to alter `renderer/+onRenderHtml.jsx` to alter the pageHtml to wrap everything in the by ant design provided `StyleProvider`

```jsx
  const pageHtml = ReactDOMServer.renderToString(
    <StyleProvider cache={extractStaticStyle.cache}>
      <Layout pageContext={pageContext}>
        <Page />
      </Layout>
    </StyleProvider>,
  );
```
Then, get the styles from that thing
```js
  const styles = extractStaticStyle(pageHtml).map((item) => item.style);
```

And also add them to the head
```js
  ${dangerouslySkipEscape(ReactDOMServer.renderToString(styles))}`
```

Tbh, the StyleProvider part could probably be moved into the layout and is not even needed i believe. Guess it's used for performance reasons, since it get's passed the cached.

### Testing
If you remove the styles entry from the `<head>` you can see the FOUC issue on the data fetching page, when reloading it.

**Reply by brillout** [reply]
I just pre-released [this change](https://github.com/vikejs/vike-react/commit/acd54666c4c391690224a11b8f08521e2df42dbd) as `vike-react@0.5.3-commit-acd5466`. I believe that should do the trick. Keep us updated.

**Comment by lewis-fidlers** [comment]
Yes! This seems to work perfectly @brillout, thanks!

For future reference. Getting the styles injected was pretty straightforward. In my layout file i wrapped the app in a Style provider
```jsx 
<StyleProvider cache={extractStaticStyle.cache}>
  <App />
</StyleProvider>
```

And in `+onAfterRenderHtml.ts`
```typescript
import { extractStaticStyle } from "antd-style";
import { useConfig } from "vike-react/useConfig";
import type { PageContext } from "vike/types";

export default (pageContext: PageContext) => {
  const config = useConfig();

  const pageHtml = pageContext.pageHtmlString;
  const styles = extractStaticStyle(pageHtml).map((item) => item.style);

  config({ Head: styles });
};

```

Which stops the The page from flashing when reloading.

**Reply by brillout** [reply]
Neat :+1: Up for publishing an example? I'll then create `vike.dev/antd`.

**Reply by lewis-fidlers** [reply]
Will do, I'll try to get it to you over the weekend. I'll adapt the overall layout as well to make it a bit more like the entire app uses ant design.

**Reply by brillout** [reply]
:+1: FYI you can also use https://vike.dev/Wrapper which would be slightly more idiomatic.

**Reply by lewis-fidlers** [reply]
I've setup an example over [here](https://github.com/lewis-fidlers/vike-react-antd/tree/main) It has an ant design version of the default "todo, star wars and counter example". 

Also used the Wrapper component as suggested. Let me know if you'd like something else added or if something is missing. 

Cheers, Lewis

**Reply by brillout** [reply]
Thank you! I've added your example to the docs https://vike.dev/antd. Also, a Vike extension `vike-react-antd` would be neat in case that's something you'd be up for. You can have a look at https://github.com/vikejs/vike-react/tree/main/packages/vike-react-apollo and https://github.com/vikejs/vike-react/tree/main/packages/vike-react-query &mdash; it's relatively simple.
